learn2track/models/gru_mixture.py

```diff
@@ -276,14 +280,65 @@ def _compute_losses(self, model_output):
+
+        # loss_per_time_step.shape = (batch_size, seq_len)
+        self.loss_per_time_step = T.sqrt(T.sum(((self.samples - self.dataset.symb_targets)**2), axis=2))
+        # loss_per_seq.shape = (batch_size,)
+        self.loss_per_seq = T.sum(self.loss_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
+
+        return self.loss_per_seq
+
+    def _get_mixture_parameters(self, regression_output):
+        batch_size, seq_len = regression_output.shape[:2]
+
+        gaussian_3d_params_shape = (batch_size, seq_len, self.n, 3)
+
+        mixture_weights = softmax(regression_output[:, :, :self.n], axis=2)
+        means = T.reshape(regression_output[:, :, self.n:self.n * 4], gaussian_3d_params_shape)
+        stds = T.reshape(T.exp(regression_output[:, :, self.n * 4:self.n * 7]), gaussian_3d_params_shape)
+
+        return mixture_weights, means, stds
 
-        # L2_errors_per_time_step.shape = (batch_size, seq_len)
-        self.L2_errors_per_time_step = T.sqrt(T.sum(((expected_value - self.dataset.symb_targets)**2), axis=2))
-        # avg_L2_error_per_seq.shape = (batch_size,)
-        self.avg_L2_error_per_seq = T.sum(self.L2_errors_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
 
-        return self.avg_L2_error_per_seq
+class MultivariateGaussianMixtureMaxComponentL2Distance(Loss):
+    """ Computes the L2 distance for the max component samples of a multivariate gaussian mixture
+    """
+
+    def __init__(self, model, dataset):
+        super().__init__(model, dataset)
+        self.n = model.n_gaussians
+        self.d = model.output_size
+
+    def _get_updates(self):
+        return {}  # There is no updates for L2Distance.
+
+    def _compute_losses(self, model_output):
+        mask = self.dataset.symb_mask
+
+        # regression_outputs.shape = (batch_size, seq_length, regression_layer_size)
+        regression_outputs = model_output
+
+        # mixture_weights.shape : (batch_size, seq_len, n_gaussians)
+        # means.shape : (batch_size, seq_len, n_gaussians, 3)
+        mixture_weights, means, stds = self._get_mixture_parameters(regression_outputs)
+
+        batch_size = mixture_weights.shape[0]
+        seq_len = mixture_weights.shape[1]
+
+        row_ids = T.tile(T.arange(0, batch_size), (seq_len,1)).T    # [[0,0,...,0],[1,1,...,1],...]
+        column_ids = T.tile(T.arange(0, seq_len), (batch_size, 1))  # [[0,1,...,seq_len],[0,1,...,seq_len],...]
+        choices = T.argmax(mixture_weights, axis=2)
+
+        # samples.shape : (batch_size, seq_len, 3)
+        self.samples = means[row_ids, column_ids, choices]
+
+        # loss_per_time_step.shape = (batch_size, seq_len)
+        self.loss_per_time_step = T.sqrt(T.sum(((self.samples - self.dataset.symb_targets)**2), axis=2))
+        # loss_per_seq.shape = (batch_size,)
+        self.loss_per_seq = T.sum(self.loss_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
+
+        return self.loss_per_seq
 
     def _get_mixture_parameters(self, regression_output):
```

Comment:
Not needed. We should use the model's method (see related comment about a generic version of that method in the GRUMixture class).